### ISSUE TITLE
Cannot delete the main window

### ISSUE THREAD

**Issue by Horrih** [emacs-lsp/lsp-treemacs#122]
Hello,
I have noticed an issue with closing windows after calling lsp-treemacs-errors-list : i cannot close the main window anymore from a secondary window. 

To reproduce : emacs 27.1, a pretty standard init.el with treemacs, lsp, lsp-treemacs used on a cpp file with clangd as backend. Dunno if it's important but I have not enabled the treemacs view on the left on startup.

1) emacs -nw main.cpp
2) M-x lsp-treemacs-errors-list
3) M-x other-window                      ;; Switch back to main.cpp
4) M-x split-window-right              ;; Open a new window on the right on main.cpp
5) M-x other-window                     ;; Switch to the right window
6) M-x delete-other-windows        ;; Try to delete the left window : does not work

Step 6 closes the treemacs window but not the left window.
If I run delete-other-windows from the left window , the right window gets closed correctly though

After some investigation, this is due to a dangling no-delete-other-windows parameter that seems set to the wrong window. I had a quick look to the source code but you don't seem to set the flag directly. I currently don't have the bandwidth to investigate more and submit a PR, but maybe in a few weeks if you don't have time on your side.

**Comment by noorul** [comment]
Any work around for this?

**Comment by jasalt** [comment]
Affects 29.1 still. @Horrih workaround is seems like best solution for now.

Two possibly related unanswered Emacs StackOverflow threads:
- https://emacs.stackexchange.com/questions/76595/help-delete-other-windows-only-working-from-the-original-window
- https://emacs.stackexchange.com/questions/77201/lsp-modes-magic-to-prevent-c-x-1-from-working

**Comment by hansbogert** [comment]
kinda irritating, but killing the window, not the buffer, seems to help.

**Comment by ryanobjc** [comment]
I ran into this problem and did a bit of debugging, and this is the backtrace of how no-delete-other-windows:

```
Debugger entered: nil
  (progn (debug))
  (if (and (eq param 'no-delete-other-windows) (eq value t)) (progn (debug)))
  debug-window-param-setter(#<window 427 on api/main.go> no-delete-other-windows t)
  apply(debug-window-param-setter (#<window 427 on api/main.go> no-delete-other-windows t))
  set-window-parameter(#<window 427 on api/main.go> no-delete-other-windows t)
  #f(compiled-function () (interactive nil) #<bytecode -0x1cbfa55d80cc94b6>)()
  treemacs--mode-check-advice(#f(compiled-function () (interactive nil) #<bytecode -0x1cbfa55d80cc94b6>))
  apply(treemacs--mode-check-advice #f(compiled-function () (interactive nil) #<bytecode -0x1cbfa55d80cc94b6>) nil)
  treemacs-mode()
  lsp-treemacs-render(...)
  lsp-treemacs-errors-list--refresh()
  run-hooks(lsp-diagnostics-updated-hook)
  apply(run-hooks lsp-diagnostics-updated-hook)
  lsp--on-diagnostics(...)
  lsp--on-notification(...)
  lsp--parser-on-message(...)
  #f(compiled-function (proc input) #<bytecode 0x1184357539949e32>)(#<process gopls<3>> ...)
```

As we can see, I once upon a time opened up the lsp-errors-list and now on an lsp event, we call lsp-treemacs-render() which ends up setting the no-delete-other-windows to the 'main' window, not one of the 'side' windows.

I'm not really sure why this is happening, I had a look at the 'treemacs--mode-check-advice' and I need to do a bit more debug.

Basically looks like a bug in treemacs, or at least how lsp uses treemacs.  I don't use treemacs normally, so maybe I can disable the lsp-treemacs integration and prevent this nonsense from happening.

**Comment by ryanobjc** [comment]
Okay I did more debugging, this is what happens:

- LSP has an event, and runs the hooks lsp-diagnostics-updated-hook
- The hook lsp-treemacs-errors-list--refresh runs
- Inside this hook it calls lsp-treemacs-render giving the buffer *LSP Error List*
- lsp-treemacs-render calls the macro (treemacs-initialize)
- This macro calls treemacs-mode
- Also treemacs--mode-check-advice is called, but it doesnt contribute much to this sequence
- Inside the treemacs-mode code, it calls this:
```
  (set-window-parameter (selected-window) 'no-delete-other-windows treemacs-no-delete-other-windows)
```

So when the I print the current buffer inside the `treemacs--mode-check-advice`, the (current-buffer) is the *LSP Error List* buffer, but the (get-buffer-window (current-buffer)) is nil.

So treemacs-mode just blindly calls set-window-parameter, assuming that the current buffer has a window and that window is (selected-window) which is not true in this case.

**Comment by ryanobjc** [comment]
This appears to be a treemacs bug and I proposed a simple solution in: https://github.com/Alexander-Miller/treemacs/issues/1149

My initial simple testing indicates that it's working great so far!

**Comment by ryanobjc** [comment]
Pull Request: https://github.com/Alexander-Miller/treemacs/pull/1150
